### ISSUE TITLE
Use Painless to compute delays on the fly

### ISSUE THREAD

**Issue by alpgarcia** [chaoss/grimoirelab-sigils#95]
Currently we are pre-computing delays. This introduces a point of failure to some panels and also makes data not as fresh as it could.

We have tested a short painless code to compute these delays on the fly that could be included in our Kibana index patterns when needed:
```
LocalDateTime creation = LocalDateTime.ofInstant(Instant.ofEpochMilli(doc['grimoire_creation_date'].value), ZoneId.of('Z'));

LocalDateTime closed;

if (doc.containsKey('state')) {
  if (doc['state'].value == 'closed') {
     closed = LocalDateTime.ofInstant(Instant.ofEpochMilli(doc['closed_at'].value), ZoneId.of('Z'));
  } else {
     Date now = new Date();
     closed = LocalDateTime.ofInstant(Instant.ofEpochMilli(now.getTime()), ZoneId.of('Z'));
  }

  return Duration.between(creation, closed).toDays();
} else {
  return 0;
}
```

**IMPORTANT**: This code has been tested and solves the problem, however some tests are needed to guarantee its performance when dealing with big indexes before including it in production.

**Closing condition**: when tests are done and, in case of good performance, index patterns are updated to use it. In the case we are unable to get good performance (for instance, not using variables [1]), discard this option.

[1] This code contains variables. It may be a bad idea because the script would be compiled each time the field needs to be solved. Number of compilations per minute is limited to 15 by default, so maybe removing `closed` and `creation` vars could improve performance if needed. See:
https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-scripting-using.html#prefer-params

**Comment by dlumbrer** [comment]
**IMPORTANT**: There are changes in the script due to type errors (see https://discuss.elastic.co/t/exception-in-kibana-scripted-field/132801/2).

The script for Jira is now:

```
if (doc.containsKey('status')) {
  if (doc['status'].value == 'Closed') {
     return Duration.between(LocalDateTime.ofInstant(Instant.ofEpochMilli(doc['grimoire_creation_date'].value.millis), ZoneId.of('Z')), LocalDateTime.ofInstant(Instant.ofEpochMilli(doc['resolution_date'].value.millis), ZoneId.of('Z'))).toMinutes()/1440.0;
  } else {
     return Duration.between(LocalDateTime.ofInstant(Instant.ofEpochMilli(doc['grimoire_creation_date'].value.millis), ZoneId.of('Z')), LocalDateTime.ofInstant(Instant.ofEpochMilli(new Date().getTime()), ZoneId.of('Z'))).toMinutes()/1440.0;
  }  
} else {
  return 0;
}
```

In github_issues:

```
if (doc.containsKey('state')) {
  if (doc['state'].value == 'closed') {
     return Duration.between(LocalDateTime.ofInstant(Instant.ofEpochMilli(doc['grimoire_creation_date'].value.millis), ZoneId.of('Z')), LocalDateTime.ofInstant(Instant.ofEpochMilli(doc['closed_at'].value.millis), ZoneId.of('Z'))).toMinutes()/1440.0;
  } else {
     return Duration.between(LocalDateTime.ofInstant(Instant.ofEpochMilli(doc['grimoire_creation_date'].value.millis), ZoneId.of('Z')), LocalDateTime.ofInstant(Instant.ofEpochMilli(new Date().getTime()), ZoneId.of('Z'))).toMinutes()/1440.0;
  }

  
} else {
  return 0;
}
```

and the field `state` must have `fielddata=true`


In Gerrit:

```
if (doc.containsKey('status')) {
  if (doc['status'].value == 'closed') {
     return Duration.between(LocalDateTime.ofInstant(Instant.ofEpochMilli(doc['grimoire_creation_date'].value.millis), ZoneId.of('Z')), LocalDateTime.ofInstant(Instant.ofEpochMilli(doc['resolution_date'].value.millis), ZoneId.of('Z'))).toMinutes()/1440.0;
  } else {
     return Duration.between(LocalDateTime.ofInstant(Instant.ofEpochMilli(doc['grimoire_creation_date'].value.millis), ZoneId.of('Z')), LocalDateTime.ofInstant(Instant.ofEpochMilli(new Date().getTime()), ZoneId.of('Z'))).toMinutes()/1440.0;
  }

  
} else {
  return 0;
}
```
and the field `status` must have `fielddata=true`

---

The scripts works as expected, we have to try them in a "real" environment with data in order to monitor them.

**Comment by dlumbrer** [comment]
The painless scripts are now updated with the new fields (https://github.com/chaoss/grimoirelab-elk/pull/566):

Jira:
```
if (doc.containsKey('status')) {
  if (doc['status'].value == 'Resolved') {
     return Duration.between(LocalDateTime.ofInstant(Instant.ofEpochMilli(doc['grimoire_creation_date'].value.millis), ZoneId.of('Z')), LocalDateTime.ofInstant(Instant.ofEpochMilli(doc['updated'].value.millis), ZoneId.of('Z'))).toMinutes()/1440.0;
  } else {
     return Duration.between(LocalDateTime.ofInstant(Instant.ofEpochMilli(doc['grimoire_creation_date'].value.millis), ZoneId.of('Z')), LocalDateTime.ofInstant(Instant.ofEpochMilli(new Date().getTime()), ZoneId.of('Z'))).toMinutes()/1440.0;
  }  
} else {
  return 0;
}
```

Gerrit:

```
if (doc.containsKey('status')) {
  if (doc['status'].value == 'closed') {
     return Duration.between(LocalDateTime.ofInstant(Instant.ofEpochMilli(doc['grimoire_creation_date'].value.millis), ZoneId.of('Z')), LocalDateTime.ofInstant(Instant.ofEpochMilli(doc['last_updated'].value.millis), ZoneId.of('Z'))).toMinutes()/1440.0;
  } else {
     return Duration.between(LocalDateTime.ofInstant(Instant.ofEpochMilli(doc['grimoire_creation_date'].value.millis), ZoneId.of('Z')), LocalDateTime.ofInstant(Instant.ofEpochMilli(new Date().getTime()), ZoneId.of('Z'))).toMinutes()/1440.0;
  }

  
} else {
  return 0;
}

```

I will change the panels and do the PR to sigils

**Comment by dlumbrer** [comment]
> The painless scripts are now updated with the new fields ([chaoss/grimoirelab-elk#566](https://github.com/chaoss/grimoirelab-elk/pull/566)):
> 
> Jira:
> 
> ```
> if (doc.containsKey('status')) {
>   if (doc['status'].value == 'Resolved') {
>      return Duration.between(LocalDateTime.ofInstant(Instant.ofEpochMilli(doc['grimoire_creation_date'].value.millis), ZoneId.of('Z')), LocalDateTime.ofInstant(Instant.ofEpochMilli(doc['updated'].value.millis), ZoneId.of('Z'))).toMinutes()/1440.0;
>   } else {
>      return Duration.between(LocalDateTime.ofInstant(Instant.ofEpochMilli(doc['grimoire_creation_date'].value.millis), ZoneId.of('Z')), LocalDateTime.ofInstant(Instant.ofEpochMilli(new Date().getTime()), ZoneId.of('Z'))).toMinutes()/1440.0;
>   }  
> } else {
>   return 0;
> }
> ```
> Gerrit:
> 
> ```
> if (doc.containsKey('status')) {
>   if (doc['status'].value == 'MERGED') {
>      return Duration.between(LocalDateTime.ofInstant(Instant.ofEpochMilli(doc['grimoire_creation_date'].value.millis), ZoneId.of('Z')), LocalDateTime.ofInstant(Instant.ofEpochMilli(doc['last_updated'].value.millis), ZoneId.of('Z'))).toMinutes()/1440.0;
>   } else {
>      return Duration.between(LocalDateTime.ofInstant(Instant.ofEpochMilli(doc['grimoire_creation_date'].value.millis), ZoneId.of('Z')), LocalDateTime.ofInstant(Instant.ofEpochMilli(new Date().getTime()), ZoneId.of('Z'))).toMinutes()/1440.0;
>   }
> 
>   
> } else {
>   return 0;
> }
> ```
> I will change the panels and do the PR to sigils

I had to add also the status `ABANDONED` in gerrit, so the new script for gerrit is:

```
if (doc.containsKey('status')) {
  if (doc['status'].value == 'MERGED' || doc['status'].value == 'ABANDONED') {
     return Duration.between(LocalDateTime.ofInstant(Instant.ofEpochMilli(doc['grimoire_creation_date'].value.millis), ZoneId.of('Z')), LocalDateTime.ofInstant(Instant.ofEpochMilli(doc['last_updated'].value.millis), ZoneId.of('Z'))).toMinutes()/1440.0;
  } else {
     return Duration.between(LocalDateTime.ofInstant(Instant.ofEpochMilli(doc['grimoire_creation_date'].value.millis), ZoneId.of('Z')), LocalDateTime.ofInstant(Instant.ofEpochMilli(new Date().getTime()), ZoneId.of('Z'))).toMinutes()/1440.0;
  }

  
} else {
  return 0;
}
```

**Comment by dlumbrer** [comment]
I had to modify the script for GitHub, cause the field `time_to_close_days` is `-` when is opened, so the new script field for GitHub Issues is:

```
if (doc.containsKey('state')) {
  if (doc['state'].value == 'closed') {
     return Duration.between(LocalDateTime.ofInstant(Instant.ofEpochMilli(doc['grimoire_creation_date'].value.millis), ZoneId.of('Z')), LocalDateTime.ofInstant(Instant.ofEpochMilli(doc['closed_at'].value.millis), ZoneId.of('Z'))).toMinutes()/1440.0;
  }
} else {
  return 0;
}
```

**Comment by dlumbrer** [comment]
> I had to modify the script for GitHub, cause the field `time_to_close_days` is `-` when is opened, so the new script field for GitHub Issues is:
> 
> ```
> if (doc.containsKey('state')) {
>   if (doc['state'].value == 'closed') {
>      return Duration.between(LocalDateTime.ofInstant(Instant.ofEpochMilli(doc['grimoire_creation_date'].value.millis), ZoneId.of('Z')), LocalDateTime.ofInstant(Instant.ofEpochMilli(doc['closed_at'].value.millis), ZoneId.of('Z'))).toMinutes()/1440.0;
>   }
> } else {
>   return 0;
> }
> ```

I was wrong, the script needs to have the else in order to compute `time_open_days`:

```
if (doc.containsKey('state')) {
  if (doc['state'].value == 'closed') {
     return Duration.between(LocalDateTime.ofInstant(Instant.ofEpochMilli(doc['grimoire_creation_date'].value.millis), ZoneId.of('Z')), LocalDateTime.ofInstant(Instant.ofEpochMilli(doc['closed_at'].value.millis), ZoneId.of('Z'))).toMinutes()/1440.0;
  } else {
     return Duration.between(LocalDateTime.ofInstant(Instant.ofEpochMilli(doc['grimoire_creation_date'].value.millis), ZoneId.of('Z')), LocalDateTime.ofInstant(Instant.ofEpochMilli(new Date().getTime()), ZoneId.of('Z'))).toMinutes()/1440.0;
  }

  
} else {
  return 0;
}
```

**Comment by dlumbrer** [comment]
I modify the Jira painless field, adding the new one just called painless_time_to_now. This field just computes the time of the issues when opened to now, in order to compute the time of just open or in progress issues on the fly. So I changed the field and now is this code:

```
if (doc.containsKey('status')) {
  return Duration.between(LocalDateTime.ofInstant(Instant.ofEpochMilli(doc['grimoire_creation_date'].value.millis), ZoneId.of('Z')), LocalDateTime.ofInstant(Instant.ofEpochMilli(new Date().getTime()), ZoneId.of('Z'))).toMinutes()/1440.0;
} else {
  return 0;
}
```

I had to modify the visualizations (from Jira Backlog panel) that used the saved search "Open-Jira-Issues-Search:_status:open_or_status:in_progress" adding this new field instead time_to_close_days and I add this saved search to some visualizations of the Jira_Timing panel.
